lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/TaskManager.cs

```diff
@@ -0,0 +1,526 @@
+        internal void SetFailedTaskInShuttingDown(IFailedTask failedTask)
+        {
+            if (TaskState(failedTask.Id) == StateMachine.TaskState.TaskWaitingForClose)
+            {
+                UpdateState(failedTask.Id, TaskStateEvent.ClosedTask);
+            }
+            else if (TaskState(failedTask.Id) != StateMachine.TaskState.TaskFailedByEvaluatorFailure)
+            {
+                UpdateState(failedTask.Id, GetTaskErrorEvent(failedTask));
+            }
+        }
+
+        /// <summary>
+        /// This method is called when receiving an IFailedEvaluator event during TaskSubmitted, TaskRunning or system ShuttingDown.
+        /// Removes the task from RunningTasks if the task associated with the FailedEvaluator is present and running. 
+        /// Sets the task state to TaskFailedByEvaluatorFailure 
+        /// </summary>
+        /// <param name="failedEvaluator"></param>
+        internal void SetTaskFailByEvaluator(IFailedEvaluator failedEvaluator)
+        {
+            if (failedEvaluator.FailedTask.IsPresent())
+            {
+                var taskId = failedEvaluator.FailedTask.Value.Id;
+                if (TaskState(taskId) == StateMachine.TaskState.TaskRunning)
+                {
+                    RemoveRunningTask(taskId);
+                }
+
+                if (TaskState(taskId) == StateMachine.TaskState.TaskWaitingForClose)
+                {
+                    UpdateState(taskId, TaskStateEvent.ClosedTask);
+                }
+                else
+                {
+                    UpdateState(taskId, TaskStateEvent.FailedTaskEvaluatorError);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Removes a task from running tasks if it exists in the running tasks collection
+        /// </summary>
+        /// <param name="taskId"></param>
+        private void RemoveRunningTask(string taskId)
+        {
+            if (_runningTasks.ContainsKey(taskId))
+            {
+                _runningTasks.Remove(taskId);
+            }
+        }
+
+        /// <summary>
+        /// Updates task state for a given taskId based on the task event
+        /// </summary>
+        /// <param name="taskId"></param>
+        /// <param name="taskEvent"></param>
+        private void UpdateState(string taskId, TaskStateEvent taskEvent)
+        {
+            if (!_tasks.ContainsKey(taskId))
+            {
+                var msg = string.Format(CultureInfo.InvariantCulture, "The task [{0}] doesn't exist.", taskId);
+                Exceptions.Throw(new IMRUSystemException(msg), Logger);
+            }
+            GetTaskInfo(taskId).TaskState.MoveNext(taskEvent);
+        }
+
+        /// <summary>
+        /// Checks if all the tasks are running.
+        /// </summary>
+        /// <returns></returns>
+        internal bool AreAllTasksRunning()
+        {
+            return _tasks.All(t => t.Value.TaskState.CurrentState == StateMachine.TaskState.TaskRunning) &&
```

Comment:
Why do we need && condition here. If one is true other will always be true right?